### ISSUE TITLE
x/tools/go/analysis: better error with incorrect version import 

### ISSUE THREAD

**Issue by kevinburke** [golang/go#70817]
### Go version

go1.23.3

### Output of `go env` in your module/workspace:

```shell
Not relevant
```

### What did you do?

I maintain a metrics library called "github.com/segmentio/stats" which is currently on version 5. I accidentally added a new file that imported "github.com/segmentio/stats" instead of "github.com/segmentio/stats/v5". Check out this branch in order to reproduce the mistake I made: https://github.com/segmentio/stats/compare/main...broken-library-version?expand=1

Golangci-lint is a popular Go code analysis tool. This tool has a "typecheck" runner, that uses x/tools/go/analysis to analyze source code: https://github.com/golangci/golangci-lint/blob/master/pkg/golinters/typecheck.go

### What did you see happen?

When you run golangci-lint with default settings (plus `--verbose`), the typecheck analyzer fails and reports the following errors: 

```go
INFO [runner/max_same_issues] 483/486 issues with text "undefined: stats" were hidden, use --max-same-issues

cmd/dogstatsd/main.go:92:3: undefined: stats (typecheck)
		stats.Add(name, value, tags...)
		^
cmd/dogstatsd/main.go:95:3: undefined: stats (typecheck)
		stats.Set(name, value, tags...)
		^
cmd/dogstatsd/main.go:163:13: undefined: stats (typecheck)
type tags []stats.Tag
```

### What did you expect to see?

The analyzer is complaining about an import "github.com/segmentio/stats/v5" and subsequent reference in the code to "stats". It's unclear why this only happens when you add an import of e.g. "github.com/segmentio/stats". While the actual problem becomes more clear if you run e.g. "go mod tidy", you need to know to run 'go mod tidy'.

This is a pretty confusing error message for end users. The code builds just fine if you run it with e.g. "go build" - it understands that the package at "github.com/segmentio/stats/v5" is named "stats." It also does not provide any context as to what the problem actually is. 

I'm pretty confident the underlying issue is in the error message generated by the analyzer, and not with golangci-lint itself, although it's definitely possible they're running the analyzer in an unexpected way. 

However, given the popularity of golangci-lint, it would substantially benefit the ecosystem if we could present useful error messages that help people solve problems, even if they are configuring the analyzer in an unexpected way.

**Comment by cagedmantis** [comment]
cc [@timothy-king](https://github.com/timothy-king)

**Comment by timothy-king** [comment]
I am going to close this as it is not actionable on our side yet.

This is really a feature request for `golangci-lint`. In particular the package, `github.com/golangci/golangci-lint/pkg/goanalysis`. I think you should file a feature request with them. If they want to implement this feature but cannot implement it with what is made available in x/tools/go/{analysis,packages} and maybe `go list` (and maybe some packages that handle modules), they can file more specific feature requests for what they need from x/tools. That would be actionable.

**Comment by kevinburke** [comment]
Thanks for investigating. I would hope that the analysis package in x/tools could provide a clearer error message or suggest things that might be wrong. As it stands it makes no sense - it’s complaining about things that compile correctly in unrelated packages to the source of the problem.

**Comment by adonovan** [comment]
If I understand correctly, the goanalysis plugin of golangci-lint is merely surfacing the errors produced by the type checker as analysis diagnostics. (I don't understand why this should be necessary since the analysis driver should be reporting type errors anyway as they would prevent most analyzers from being able to run; but perhaps golangci-lint does things differently for some reason.) If so, then this is actually a complaint about the error messages produced by the type checker (which are very similar to the ones produced by cmd/compile front-end because they share most of their logic).

@kevinburke, could you confirm whether this is so? If so, I would expect it to be possible to produce examples of confusing errors using only the Go playground.

**Comment by kevinburke** [comment]
Yes - my concern is the error messages are confusing because they are flagging files that still compile successfully when you run "go build", and don't provide any information that can be used to diagnose the actual problem.

**Comment by kevinburke** [comment]
OK. Would you mind if we could leave this issue open then until I have time to figure out where the issue actually lies. 

Alternatively, are you saying the issue is that golangci-lint should be using something else to typecheck besides the analysis frontend?

**Comment by thepudds** [comment]
Hi @kevinburke, if you are interested in triaging this, one approach could be to use the fairly convenient 
[x/tools/go/analysis/singlechecker](https://pkg.go.dev/golang.org/x/tools/go/analysis/singlechecker) as a sanity check.

For example, I somewhat arbitrarily picked the ifaceassert analyzer from x/tools and configured singlechecker to use it ([playground link](https://go.dev/play/p/6OF9x7_ptgw)) and then ran it locally on Alan's example from above. It printed this error, which includes the package path (which is similar to Alan's result above from cmd/compile's error message):

> alan-example/main.go:8:25: cannot use new(htmpl.Template) (value of type *"html/template".Template) as *"text/template".Template value in variable declaration

There is also the new [x/tools/go/analysis/checker](https://pkg.go.dev/golang.org/x/tools/go/analysis/checker), which is maybe similar in spirit to singlechecker but with more control for power users.

In any event, probably worth opening an issue with golangci-lint and/or seeing if it's possible to reproduce the bad error with a simpler reproducer.
